Merge secrets from command line options and YAML file

ericstoekl opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
@alexellis mentioned in the Slack channel that we should merge the `secrets` from the command line flags with the `secrets` from the stack YAML. Currently `faas-cli` will only take the secrets from the YAML file **or** the command line flags. For example, this command:

`faas-cli deploy -f stack.yml --secret api_token_2`

Where `stack.yml` contains:

```
  functions:
    fun1:
      image: alpine:latest
      fprocess: cat
      secret: api_token_1
```

In the current state of the code, only `api_token_1` will be taken, because the secret passed in as a command line flag (`api_token_2`) will be ignored.

## Description
<!--- Describe your changes in detail -->
This change adds a function which merges all of the secrets for the function found in the stack YAML with all of the secrets found from the command line flags.

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
- [X] I have raised an issue to propose this change ([required](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md))
#227 

This issue was partially resolved by this merge: #235 

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
Tested using the instructions and bash script file in this repo: https://github.com/ericstoekl/secrets-yaml-test

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] My code follows the code style of this project.
- [X] My change requires a change to the documentation.
- [X] I have updated the documentation accordingly.
- [X] I've read the [CONTRIBUTION](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md) guide
- [X] I have signed-off my commits with `git commit -s`
- [X] I have added tests to cover my changes.
- [X] All new and existing tests passed.

alexellis commented:
Great

alexellis commented on commands/deploy.go:
```diff
@@ -326,3 +324,14 @@ func deriveFprocess(function stack.Function) (string, error) {
 }
+
+func compileSecrets(cmdLineSecrets []string, yamlSecrets []string) []string {
+	var merged []string
```
We should make sure we first take all the YAML Then overlay (replace) any keys found in the CLI / cmd-line args. We don't want to have duplicates/conflicts.

ericstoekl commented:
@alexellis , Thanks for the comments. I've updated the PR in accordance with the requested changes.

I've updated my test script to also check the number of secrets that exist on the container once it's created, to make sure that duplicate secrets (passing the same secret value on the command line that already exists in the stack YAML) only show up as one secret in the container:

https://github.com/ericstoekl/secrets-yaml-test

alexellis commented on commands/deploy.go:
```diff
@@ -326,3 +324,26 @@ func deriveFprocess(function stack.Function) (string, error) {
 }
+
+func compileSecrets(yamlSecrets []string, cmdLineSecrets []string) []string {
+	secretsMap := make(map[string]bool)
```
I was thinking something like what was done for labels/env-vars? Can this be used? https://github.com/ericstoekl/faas-cli/blob/e5d279cfdde669fbc6a51ad42fe003a8f2d02ed4/commands/deploy.go#L264

ericstoekl commented:
@alexellis Thanks for your comments, I've updated again, and squashed the commits. Here's what it does now in the `compileSecrets()` function:

1. Secrets come in from the YAML file and from the Command Line, both as type `[]string`. I convert each array to a `map[string]string`.
2. Send both `map`s to the `mergeMap()` function. The result is returned in the form of a `map[string]string`.
3. Convert the result of `mergeMap()` back into a `[]string`.

Let me know if this is acceptable. I've tested using the aforementioned bash test script.

Thanks,
Eric

alexellis commented:
Hey I appreciate you working on this. I took a few minutes to review it again and thought it might be easier to raise a separate PR. Closing this one in favour of #241
